File: python/VPIAlgo.py
from utils import *
#from scipy.special import softmax
from stateEliminationAlgo import *

class VPIAlgo(StateEliminationAlgo):
    def generate_proposals(self):
        # compute all possible proposals that this agent can give
        for agent in self.game.agents:
            agent.proposals = generate_proposals(self.game.state, agent)

    def evaluate_coalition(self, coalition):
        W = [agent.agent_type for agent in coalition]
        return eval_coalition(W, self.game.tasks, False)
    @jit
    def softmax(self, x):
        # https://stackoverflow.com/questions/34968722/how-to-implement-the-softmax-function-in-python
        """Compute softmax values for each sets of scores in x."""
        e_x = np.exp(x - np.max(x))
        return e_x / e_x.sum(axis=0)  # only difference

    def calculate_VPIs(self, proposer, best_proposals, best_value, second_best, proposals):
        VPIs = [] # value of perfect information of each of the proposals
        for proposal in proposals:
            coalition, div = proposal
            proposer_share =div[coalition.index(proposer)]
            VPI = 0
            for agent_types in product(range(1, self.game.T + 1), repeat=len(coalition)):
                prob = reduce(mul, \
                              [proposer.belief[agent][other_agent_type - 1] for \
                               agent, other_agent_type in zip(coalition, agent_types)])
                predicted_reward = proposer_share * eval_coalition(list(agent_types), self.game.tasks)
                if proposal in best_proposals and predicted_reward < second_best:
                    VPI += prob * (second_best - predicted_reward)
                elif predicted_reward > best_value:
                    VPI += prob * (predicted_reward - best_value)
            VPIs.append(VPI)
        return VPIs
    def proposal_outcome(self):
        # one-step proposer process using informed belief
        # only return feasible coalition!! Singleton for the proposer
        # is always feasible!
        proposer = nature_pick_proposer(self.game.state)


        proposals = [] # list of proposals
        proposal_values = [] # corresponding list of values
        best_value = second_best = 0
        best_proposals = set()
        # consider all proposals and all divisions possible!
        # choose proposal based on softmax selection to facilitate exploration
        #for proposal in proposer.proposals:
        for proposal in generate_proposals(self.game.state, proposer, True):
            coalition, div = proposal
            responses, predicted_reward, continuation_payoff = \
                predict_responses(proposer, proposal, self.game, ret_extra=True)
            disagree = [response == 'no' for player, response in responses.items() if player != proposer]
            if any(disagree):
                proposal_value = continuation_payoff[proposer]# proposal fails! the proposer get the reserve value
            else:
                proposal_value = div[coalition.index(proposer)] * predicted_reward

            #print('proposal, responses, predicted_reward, cont_payoff, proposal_val:', proposal,
            #      responses, predicted_reward, continuation_payoff, proposal_value)

            proposals.append(proposal)
            proposal_values.append(proposal_value)
            if proposal_value > best_value:
                second_best = best_value
                best_value = proposal_value
                best_proposals = set(proposal)
            elif proposal_value > second_best and\
                proposal_value < best_value:
                second_best = proposal_value
            elif proposal_value == best_value:
                best_proposals.add(proposal)

        VPIs = self.calculate_VPIs(proposer, best_proposals, best_value, second_best, proposals)

        QVs = np.array(VPIs) + np.array(proposal_value)
        probs = self.softmax(QVs)
        chosen = np.random.choice(range(len(proposals)), p=probs)
        #print('ret:', proposals[ret])
        return proposals[chosen], proposer



    def response(self, proposal):
        coalition, div = proposal
        responses = {}
        for agent in coalition:
            # each agent calculate their QVs value again restricted to only yes (current proposal)
            # and no (singleton proposal) and then use softmax selection!

            accept_reward = div[coalition.index(agent)] * expected_coalition_value(coalition, agent, self.game)
            reject_reward = expected_continuation_payoff(agent, self.game)[agent]
            #rewards = sort
            #self.calculate_VPIs(agent, best_proposals, best_value, second_best, proposals)

            # simply softmax these guys for now to avoid dealing with edge cases when two choices are already equal!
            probs = self.softmax([accept_reward, reject_reward])
            agent_response = np.random.choice(['yes', 'no'], p=probs)
            responses[agent] = agent_response
        return responses

File: C++/VPIAlgo.cpp
#include "VPIAlgo.hpp"

// select proposals based on softmax on QV values!
// TODO: memorize VPI values as well!
pair<int, Proposal> VPIAlgo::proposalOutcome()
{
    Agent &proposer = game.agents[distribution(generator)];

    //cout << ">> PROPOSER WEIGHT " << proposer.weight
    //     << " |beliefChanged? " << proposer.beliefChanged << " | bestProposal size : "
    //     << proposer.bestProposals.size() << endl;

    //cout << "PRINTING PROPOSAL SPACE:" << endl;
    //proposer.printProposalSpace();
    // only compute bestProposals again if my belief has changed!
    if (proposer.beliefChanged || proposer.proposalValues.size() == 0)
    {
        populateAgentValuesAndQVs(proposer);
    }

    //cout << ">> PROPOSER " << proposer.name << " proposalValues:" << proposer.proposalValues << endl;
    //cout << "proposalValues:" << proposer.proposalValues << endl;
    VectorXd probs = softmax(proposer.QVs);
    //cout << "softmax " << probs << endl;
    //cout << "softmax probs:" << probs << endl;
    //cout << "softmax sum :" << probs.sum() << endl;
    int chosen = selectAction(probs, generator, u);

    return make_pair(proposer.name, proposer.proposalSpace[chosen]);
}

void VPIAlgo::populateAgentValuesAndQVs(Agent &agent)
{

    int numProposals = agent.proposalSpace.size();
    double bestValue = 0;
    double secondBestValue = 0;
    set<int> bestProposals = {};

    if (agent.proposalValues.size() == 0)
    {
        agent.proposalValues.resize(numProposals);
    }

    if (agent.QVs.size() == 0)
    {
        agent.QVs.resize(numProposals);
    }

    double bestReward = numeric_limits<double>::min();
    for (auto [i, proposal] : enumerate(agent.proposalSpace))
    {
        pair<double, map<int, int>> val_and_responses = game.predictReponses(agent, proposal);
        // proposerValue = coalitionValue * proposerShare
        double proposalValue = (val_and_responses.first) * proposal.second[distance(proposal.first.begin(),
                                                                                    proposal.first.find(agent.name))];
        map<int, int> &responses = val_and_responses.second;
        // remove my response since I don't care about my response
        responses.erase(agent.name);
        // product of values of map
        // if there's any zero (any "no" response), then the product would
        // be 0.
        int agreement = accumulate(begin(responses), end(responses), 1,
                                   [](int value, const std::map<int, int>::value_type &p) { return value * p.second; });

        if (!agreement)
        {
            // get the singleton value ...
            proposalValue = game.evaluateCoalition({agent.weight});
        }
        agent.proposalValues(i) = proposalValue;

        if (proposalValue > bestValue)
        {
            secondBestValue = bestValue;
            bestValue = proposalValue;
            bestProposals = set<int>{i};
        }
        else if (proposalValue > secondBestValue and proposalValue < bestValue)
        {
            secondBestValue = proposalValue;
        }
        else if (proposalValue == bestValue)
        {
            bestProposals.insert(i);
        }
    }
    agent.QVs = agent.proposalValues + calculateVPIs(agent, bestProposals, bestValue,
                                                     secondBestValue);
}
// response according to QV values!
pair<CoalitionStructure, vector<Coalition>> VPIAlgo::formationProcess()
{
    //cout << "begin formationProc" << endl;
    // record the coalition Structure!
    CoalitionStructure CS;
    pair<int, Proposal> proposer_and_proposal = proposalOutcome();
    Proposal &proposal = proposer_and_proposal.second;
    Agent &proposer = game.agents[proposer_and_proposal.first];

    Coalition &coalition = proposal.first;
    VectorXd &div = proposal.second;
    //cout << "finish picking proposals!" << endl;
    map<int, int> responses;

    VectorXd singleShare(1);
    singleShare << 1.0;

    for (int agentName : coalition)
    {
        Agent &agent = game.agents[agentName];
        if (agent.beliefChanged || agent.proposalValues.size() == 0)
        {
            populateAgentValuesAndQVs(agent);
        }

        // the value of acceptance is as if this agent was to offer the same proposal himself
        // the value of rejection is the singleton offer
        int accept_idx = getIndex(agent.proposalSpace, proposal);
        int reject_idx = getIndex(agent.proposalSpace, make_pair(set<int>{agentName}, singleShare));

        //cout << "accept_idx " << accept_idx << " reject_idx: " << reject_idx << endl;
        //cout << "agent.QVs.size()" << agent.QVs.size() << endl;
        double accept_QV = agent.QVs(accept_idx);
        double reject_QV = agent.QVs(reject_idx);
        VectorXd accept_or_reject(2);
        accept_or_reject << accept_QV, reject_QV;
        VectorXd probs = softmax(accept_or_reject);
        int chosen = selectAction(probs, generator, u);
        //cout << "chosen:" << chosen << endl;
        if (chosen == 0)
        {
            responses[agentName] = 1; // yes!
        }
        else
        {
            responses[agentName] = 0; // no!
        }
    }
    // delete my response since I don't care (exploration in the first place!)
    responses.erase(proposer.name);
    int agreement = accumulate(begin(responses), end(responses), 1,
                               [](int value, const std::map<int, int>::value_type &p) { return value * p.second; });

    //cout << "agreement? " << agreement << endl;
    if (proposal.first.size() == 2)
    {
        cout << "proposaal size == 2, agreement? " << agreement << endl;
    }
    if (agreement)
    {
        //cout << "=>> Coalition agreement!" << endl;
        // one (potentially) non-singleton coalition
        for (auto &agent : game.agents)
        {
            // if agent is in coalition
            if (coalition.find(agent.name) != coalition.end())
            {
                continue;
            }
            // agent not found in coalition
            CS.push_back(make_tuple(set<int>{agent.name},
                                    singleShare,
                                    game.evaluateCoalition({agent.weight})));
        }
        CS.push_back(make_tuple(coalition, div, game.evaluateCoalition(coalition)));

        return make_pair(CS, vector<Coalition>{coalition});
    }
    else
    {
        // all singleton coalitions
        for (auto &agent : game.agents)
        {
            CS.push_back(make_tuple(set<int>{agent.name},
                                    singleShare,
                                    game.evaluateCoalition({agent.weight})));
        }

        return make_pair(CS, vector<Coalition>{});
    }
}

// this VPI calculation does NOT take into account rejection possibility, which might be problematic ...
// although this is probably mitigated by the fact that QV = proposalValue + VPI
VectorXd VPIAlgo::calculateVPIs(Agent &proposer, set<int> bestProposals, double bestValue,
                                double secondBestValue)
{
    VectorXd VPIs(proposer.proposalSpace.size());
    int proposalNum = -1;
    for (auto &proposal : proposer.proposalSpace)
    {
        proposalNum++;
        Coalition &coalition = proposal.first;
        VectorXd &div = proposal.second;
        // In Python, proposerShare = div[coalition.index(proposer)]
        double proposerShare = div(distance(coalition.begin(),
                                            coalition.find(proposer.name)));
        double VPI = 0;
        for (auto &weights : CartesianSelfProduct(game.weightRange, coalition.size()))
        {
            double prob = 1.0;
            for (auto [i, agentName] : enumerate(coalition))
            {
                prob *= proposer.belief(agentName, weights[i] - game.minWeight);
            }

            // TODO: change this to take into account rejection!
            double predictedReward = proposerShare * game.evaluateCoalition(weights);
            // if this proposal is in bestProposals
            if (bestProposals.find(proposalNum) != bestProposals.end() &&
                predictedReward < secondBestValue)
            {
                VPI += prob * (secondBestValue - predictedReward);
            }
            else if (predictedReward > bestValue)
            {
                VPI += prob * (predictedReward - bestValue);
            }
        }
        VPIs(proposalNum) = VPI;
    }
    return VPIs;
}